ch.sbb.scion.rcp.microfrontend/src/ch/sbb/scion/rcp/microfrontend/SciManifestService.java

```diff
@@ -30,26 +30,29 @@
 @Component(service = SciManifestService.class)
 public class SciManifestService {
 
+  private volatile CompletableFuture<List<Application>> applications;
+
   @Reference
   private MicrofrontendPlatformRcpHost microfrontendPlatformRcpHost;
 
   /**
    * @see https://scion-microfrontend-platform-api.vercel.app/classes/ManifestService.html#applications
    */
   public CompletableFuture<List<Application>> getApplications() {
-    var applications = new CompletableFuture<List<Application>>();
-    new JavaCallback(microfrontendPlatformRcpHost.whenHostBrowser, args -> {
-      applications.complete(GsonFactory.create().fromJson((String) args[0], new ParameterizedType(List.class, Application.class)));
-    })
-        .installOnce()
-        .thenAccept(callback -> {
-          new JavaScriptExecutor(microfrontendPlatformRcpHost.hostBrowser, Resources.readString("js/sci-manifest-service/lookup-applications.js"))
-              .replacePlaceholder("callback", callback.name)
-              .replacePlaceholder("refs.ManifestService", Refs.ManifestService)
-              .replacePlaceholder("helpers.toJson", Helpers.toJson)
-              .execute();
-        });
-
+    if (applications == null) {
+      applications = new CompletableFuture<List<Application>>();
+      new JavaCallback(microfrontendPlatformRcpHost.whenHostBrowser, args -> {

```

Comment:
Technically, this fetch operation could still be executed multiple times as the null check and the assignment of the applications field happen in different instructions, evidently. Ofc, there is little possibility that this actually happens and from what I can tell there is no problem if it happens. Still, what about using an AtomicBoolean to indicate whether the fetch operation was already issued, and initializing the applications field on declaration (and declaring it final)?

Reply:
But we are here in the UI thread. There is only a single UI thread. What do you think?

Reply:
Yes, that's correct. The only usages of this method that I found were in the UI thread (inside the creation logic of UI components). Agreed!

Reply:
All our services should be invoked exclusively in the UI-Thread because they interact with the SWT Browser. If invoked from another thread than the UI thread, it throws an illegal thread access exception. I think we should also guard our services that way, similarly to the `Widget.checkWidget()` method. But we can do this in a future commit.

Reply:
I have removed the `volatile` keyword. Could you please take a look at it?